File: python/ratelimiter/factory.py
from __future__ import annotations

from typing import Any

from .algorithms.base import RateLimitAlgorithm
from .ratelimiter import RateLimiter
from .registry import registry


def create_rate_limiter(
    redis_client: Any,
    *,
    algorithm: str | RateLimitAlgorithm = "token_bucket",
    script_path: str | None = None,
    key_prefix: str | None = None,
) -> RateLimiter:
    algo = registry.create(algorithm) if isinstance(algorithm, str) else algorithm
    return RateLimiter(redis_client, algorithm=algo, script_path=script_path, key_prefix=key_prefix)

File: go/ratelimiter/factory.go
package ratelimiter

// CreateRateLimiter constructs a RateLimiter using an algorithm registered by name.
func CreateRateLimiter(client RedisClient, algorithmName string, opts ...Option) (*RateLimiter, error) {
	algorithm, err := Registry.Create(algorithmName)
	if err != nil {
		return nil, err
	}
	return New(client, append([]Option{WithAlgorithm(algorithm)}, opts...)...)
}
